File: node/mods/rustscript/lib/opcodes/arrayify.js
/**
 * Purpose: ARRAYIFY opcode — replace a context value with deep clones of itself.
 *
 * Language shape:
 *   script.reference  — context.* path to clone in place
 *   script.dimension  — copy count (literal, resolve_ref, or tx.to/from.p2sh)
 *
 * Read-only destinations (script.*, witness.*, tx.*, blk.*) are rejected.
 */

module.exports = {
  name: 'ARRAYIFY',
  description: `
Replaces a context value with an array of deep copies of that value.

dimension controls how many independent clones are created. It may be a
numeric literal, a normal RustScript reference (array length, object key
count, or number), or the special collection refs tx.from / tx.to /
tx.path / tx.from.p2sh / tx.to.p2sh.

Only context.* destinations are writable.
`,
  exampleScript: {
    op: 'ARRAYIFY',
    reference: 'context.constitution',
    dimension: 'tx.to.p2sh'
  },
  schema: {
    reference: 'string',
    dimension: 'reference'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const reference = node.reference;
    if (typeof reference !== 'string' || reference.length === 0) {
      return false;
    }

    if (!Object.prototype.hasOwnProperty.call(node, 'dimension')) {
      return false;
    }

    if (!reference.startsWith('context.')) {
      return false;
    }

    const path = reference.slice('context.'.length);
    if (!path || isForbiddenWritePath(path)) {
      return false;
    }

    const dimension = resolveDimension(node.dimension, context);
    if (dimension === null) {
      return false;
    }

    const original = getContextPath(context, path);
    if (original === undefined) {
      return false;
    }

    const clones = [];
    for (let i = 0; i < dimension; i += 1) {
      clones.push(deepClone(original));
    }

    return setContextPath(context, path, clones);
  }
};

function isForbiddenWritePath(path) {
  return (
    path === 'script' ||
    path.startsWith('script.') ||
    path === 'witness' ||
    path.startsWith('witness.') ||
    path === 'tx' ||
    path.startsWith('tx.') ||
    path === 'blk' ||
    path.startsWith('blk.')
  );
}

function pathSegments(path) {
  return String(path)
    .replace(/\[/g, '.')
    .replace(/\]/g, '')
    .split('.')
    .filter((s) => s.length > 0);
}

function getContextPath(root, path) {
  const segments = pathSegments(path);
  if (!segments.length) {
    return undefined;
  }
  let current = root;
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const asIndex = /^[0-9]+$/.test(segment) ? Number(segment) : null;
    if (asIndex !== null) {
      if (!Array.isArray(current) || asIndex >= current.length) {
        return undefined;
      }
      current = current[asIndex];
      continue;
    }
    if (
      !current ||
      typeof current !== 'object' ||
      Array.isArray(current) ||
      !Object.prototype.hasOwnProperty.call(current, segment)
    ) {
      return undefined;
    }
    current = current[segment];
  }
  return current;
}

function setContextPath(root, path, value) {
  const segments = pathSegments(path);
  if (!segments.length) {
    return false;
  }

  let current = root;
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const isLast = i === segments.length - 1;
    const asIndex = /^[0-9]+$/.test(segment) ? Number(segment) : null;

    if (isLast) {
      if (asIndex !== null) {
        if (!Array.isArray(current) || asIndex >= current.length) {
          return false;
        }
        current[asIndex] = value;
        return true;
      }
      if (!current || typeof current !== 'object' || Array.isArray(current)) {
        return false;
      }
      current[segment] = value;
      return true;
    }

    if (asIndex !== null) {
      if (!Array.isArray(current) || asIndex >= current.length) {
        return false;
      }
      current = current[asIndex];
      continue;
    }

    if (!current || typeof current !== 'object' || Array.isArray(current)) {
      return false;
    }
    if (!Object.prototype.hasOwnProperty.call(current, segment)) {
      current[segment] = {};
    }
    current = current[segment];
  }

  return false;
}

function dimensionFromValue(value) {
  if (typeof value === 'number' && Number.isFinite(value) && value >= 0) {
    return Math.floor(value);
  }
  if (Array.isArray(value)) {
    return value.length;
  }
  if (value && typeof value === 'object') {
    return Object.keys(value).length;
  }
  return null;
}

function countP2shSlips(slips) {
  if (!Array.isArray(slips)) {
    return 0;
  }
  let count = 0;
  for (let i = 0; i < slips.length; i += 1) {
    const slip = slips[i];
    if (!slip || typeof slip !== 'object') {
      continue;
    }
    // Mirror Rust: skip Bound; P2SH custody marked by public_key[0] == 0x00.
    if (slip.slip_type === 'Bound' || slip.slip_type === 9) {
      continue;
    }
    const pk = slip.public_key || slip.publicKey;
    if (typeof pk === 'string' && pk.length > 0) {
      // base58 keys are opaque here; editor path may pass hex prefix bytes
      continue;
    }
    if (Array.isArray(pk) && pk[0] === 0) {
      count += 1;
      continue;
    }
    if (pk && typeof pk === 'object' && pk[0] === 0) {
      count += 1;
    }
  }
  return count;
}

function resolveDimension(dimension, context) {
  const resolved = resolveRef(context, dimension);
  const fromResolved = dimensionFromValue(resolved);
  if (fromResolved !== null) {
    return fromResolved;
  }

  if (typeof dimension !== 'string') {
    return null;
  }

  if (dimension === 'tx.from') {
    return context.tx && Array.isArray(context.tx.from) ? context.tx.from.length : null;
  }
  if (dimension === 'tx.to') {
    return context.tx && Array.isArray(context.tx.to) ? context.tx.to.length : null;
  }
  if (dimension === 'tx.path') {
    return context.tx && Array.isArray(context.tx.path) ? context.tx.path.length : null;
  }
  if (dimension === 'tx.to.p2sh') {
    return context.tx ? countP2shSlips(context.tx.to) : null;
  }
  if (dimension === 'tx.from.p2sh') {
    return context.tx ? countP2shSlips(context.tx.from) : null;
  }

  return null;
}

function deepClone(value) {
  return JSON.parse(JSON.stringify(value));
}

function resolveRef(root, ref) {
  if (typeof ref !== 'string') {
    return ref;
  }
  const parts = ref.split('.');
  let cursor = root;
  for (let i = 0; i < parts.length; i += 1) {
    const key = parts[i];
    if (
      !cursor ||
      typeof cursor !== 'object' ||
      !Object.prototype.hasOwnProperty.call(cursor, key)
    ) {
      return ref;
    }
    cursor = cursor[key];
  }
  return cursor;
}

File: rust/saito-core/src/core/consensus/scripting/opcodes/arrayify.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::slip::{Slip, SlipType};
use crate::core::consensus::transaction::Transaction;
use serde_json::Value;

use super::super::script::resolve_ref;

/// ARRAYIFY — replace a context value with an array of deep clones of itself.
///
/// Script shape:
/// ```json
/// {
///   "op": "ARRAYIFY",
///   "reference": "context.constitution",
///   "dimension": 2
/// }
/// ```
///
/// `dimension` may be a numeric literal, a normal resolve_ref result (array
/// length, object key count, or number), or the special collection refs
/// `tx.from` / `tx.to` / `tx.path` / `tx.from.p2sh` / `tx.to.p2sh`.
///
/// Destinations must be `context.*` paths. Writes to `script.*`, `witness.*`,
/// `tx.*`, and `blk.*` are rejected.
pub struct Arrayify {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

fn is_forbidden_write_path(path: &str) -> bool {
    path == "script"
        || path.starts_with("script.")
        || path == "witness"
        || path.starts_with("witness.")
        || path == "tx"
        || path.starts_with("tx.")
        || path == "blk"
        || path.starts_with("blk.")
}

/// Normalize `foo[0].bar` into dotted segments compatible with resolve_ref lookup.
fn path_segments(path: &str) -> Vec<String> {
    let normalized = path.replace('[', ".").replace(']', "");
    normalized
        .split('.')
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

fn get_context_path(root: &Value, path: &str) -> Option<Value> {
    let segments = path_segments(path);
    if segments.is_empty() {
        return None;
    }

    let mut current = root;
    for segment in &segments {
        if let Ok(index) = segment.parse::<usize>() {
            current = current.get(index)?;
        } else {
            current = current.get(segment)?;
        }
    }
    Some(current.clone())
}

fn set_context_path(root: &mut Value, path: &str, value: Value) -> bool {
    let segments = path_segments(path);
    if segments.is_empty() {
        return false;
    }

    let last = segments.len() - 1;
    let mut current = root;

    for (i, segment) in segments.iter().enumerate() {
        if i == last {
            if let Ok(index) = segment.parse::<usize>() {
                let Some(arr) = current.as_array_mut() else {
                    return false;
                };
                if index >= arr.len() {
                    return false;
                }
                arr[index] = value;
                return true;
            }

            if !current.is_object() {
                *current = Value::Object(serde_json::Map::new());
            }
            if let Value::Object(map) = current {
                map.insert(segment.clone(), value);
                return true;
            }
            return false;
        }

        if let Ok(index) = segment.parse::<usize>() {
            let Some(arr) = current.as_array_mut() else {
                return false;
            };
            if index >= arr.len() {
                return false;
            }
            current = &mut arr[index];
            continue;
        }

        if !current.is_object() {
            *current = Value::Object(serde_json::Map::new());
        }

        let Value::Object(map) = current else {
            return false;
        };

        if !map.contains_key(segment) {
            map.insert(segment.clone(), Value::Object(serde_json::Map::new()));
        }

        current = map.get_mut(segment).unwrap();
    }

    false
}

fn dimension_from_value(value: &Value) -> Option<usize> {
    match value {
        Value::Number(n) => {
            let Some(u) = n.as_u64() else {
                return None;
            };
            usize::try_from(u).ok()
        }
        Value::Array(arr) => Some(arr.len()),
        Value::Object(map) => Some(map.len()),
        _ => None,
    }
}

/// Count P2SH slips using the same filter as resolve_p2sh_slip_field.
fn count_p2sh_slips(slips: &[Slip]) -> usize {
    let mut count = 0usize;
    for slip in slips {
        if slip.slip_type == SlipType::Bound {
            continue;
        }
        if slip.public_key[0] != 0x00 {
            continue;
        }
        count += 1;
    }
    count
}

fn resolve_dimension(
    dimension: &Value,
    context: &Value,
    tx: Option<&Transaction>,
    blk: Option<&Block>,
) -> Option<usize> {
    let resolved = resolve_ref(dimension, context, tx, blk);
    if let Some(n) = dimension_from_value(&resolved) {
        return Some(n);
    }

    //
    // special collection refs — only after normal resolve_ref fails to yield
    // a usable dimension
    //
    let Some(s) = dimension.as_str() else {
        return None;
    };

    match s {
        "tx.from" => tx.map(|t| t.from.len()),
        "tx.to" => tx.map(|t| t.to.len()),
        "tx.path" => tx.map(|t| t.path.len()),
        "tx.from.p2sh" => tx.map(|t| count_p2sh_slips(&t.from)),
        "tx.to.p2sh" => tx.map(|t| count_p2sh_slips(&t.to)),
        _ => None,
    }
}

impl Arrayify {
    pub fn validate(context: &mut Value, tx: Option<&Transaction>, blk: Option<&Block>) -> u8 {
        let reference = context["script"]["reference"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if reference.is_empty() {
            return 0;
        }

        let Some(path) = reference.strip_prefix("context.") else {
            return 0;
        };
        let path = path.to_string();
        if path.is_empty() || is_forbidden_write_path(&path) {
            return 0;
        }

        if context["script"].get("dimension").is_none() {
            return 0;
        }

        let Some(dimension) = resolve_dimension(
            &context["script"]["dimension"].clone(),
            context,
            tx,
            blk,
        ) else {
            return 0;
        };

        let Some(original) = get_context_path(context, &path) else {
            return 0;
        };

        let mut clones = Vec::with_capacity(dimension);
        for _ in 0..dimension {
            clones.push(original.clone());
        }

        if set_context_path(context, &path, Value::Array(clones)) {
            1
        } else {
            0
        }
    }
}
